### DISCUSSION TITLE
DataSink misunderstanding?

### DISCUSSION THREAD

**Discussion by klumbe** [eclipse-edc/Connector#4642] [General]
Hi,

when reading the official [documentation](https://eclipse-edc.github.io/documentation/),
I was under the impression, that a DataSink could be used to store data received by the consumers data plane.
When diving deeper into it, in particular when dealing with the transfer requests, it rather looks like a DataSink is used on the provider end to transfer the data to a specific target.
Or maybe both cases are possible?

Can someone point me into the right direction, possibly linking documentation that makes this a bit clearer?

In particular, I am interested in transferring (http/s) the data solely through the data planes on both the providers and the consumers end (data plane to data plane).
How could I achieve that the data-plane then stores it at the desired location?

I am particularly NOT interested in getting any final solution, I just want to know where my extension should be seated.

**Comment by ndr-brt** [comment]
Hi-level docs about data plane:
https://eclipse-edc.github.io/documentation/for-adopters/data-plane/

In depth documentation:
https://eclipse-edc.github.io/documentation/for-contributors/data-plane/

If there are things that are not clear or you feel like they could described better, feel free to open an issue on the [documentation repository](https://github.com/eclipse-edc/eclipse-edc.github.io)

**Reply by klumbe** [reply]
I've read them all and no, they don't answer my question.
I am looking for a way to store the data retrieved by a data plane.

**Reply by jimmarino** [reply]
Did you look at the samples and the DataSink implementations? There are plenty of examples of how to do this, e.g. the `KafkaDataSink` or `AsyncStreamingDataSink`. 

Note that you need to distinguish the `DataSink` implementation on the client data plane versus the provider data plane. The provider implementation should send the data over a remote transport (e.g. HTTP), while on the client it can send it to a stream or other destination for storage.

**Reply by klumbe** [reply]
Thanks, Jim.
Sounds like it would take care of both use cases (which makes sense).

I've been looking at the `AsyncStreamingDataSink` the last days and also thought it would be doing what I actually want, but I am missing two pieces:
1. The `Part` is passed to an output-stream. Where is it defined which particular one is used?
Not the biggest issue, since I could create my custom `transferPart` method and do it explicitely.
2. How is the decision for using that particular sink made on the data-consumer?
As far as I understand, the `dataDestination` field in the transfer-start request is used by the provider, right?
I would need a way to tell the consumer to pass the data over to that particular sink.

Appreciate any hint :-)

**Reply by jimmarino** [reply]
> Thanks, Jim. Sounds like it would take care of both use cases (which makes sense).
> 
> I've been looking at the `AsyncStreamingDataSink` the last days and also thought it would be doing what I actually want, but I am missing two pieces:
> 
> 1. The `Part` is passed to an output-stream. Where is it defined which particular one is used?
>    Not the biggest issue, since I could create my custom `transferPart` method and do it explicitely.
> 2. How is the decision for using that particular sink made on the data-consumer?
>    As far as I understand, the `dataDestination` field in the transfer-start request is used by the provider, right?
>    I would need a way to tell the consumer to pass the data over to that particular sink.
> 
> Appreciate any hint :-)

Hi,

Take a look at `DataPlanePublicApiV2Controller` which creates an `AsyncStreamingDataSink` and uses a `AsyncResponse` from Jersey to stream the data to a response. You could do something similar or stream it into a persistent store. In practice the way this would work is that your client app would trigger this transfer either via an API call or directly similar to the way the controller operates.

**Reply by klumbe** [reply]
I slightly checked the `DataPlanePublicApiV2Controller` as well, will take a deeper look, thanks.
So the idea is to create a custom implementation of it ?

I actually wanted to achieve something similar like the http-streaming example: 
Automatically push files over to the consumer once new data arrives and the consumer should then store it.

In theory, the nicest way would be to declare sort of an internal `dataDestination` that is only used by the consumer to evaluate the used `DataSink` based on the transfer process.

**Reply by jimmarino** [reply]
Ah, so you want a PUSH, not PULL transfer type.  Why not create an endpoint (controller) and have that store the data directly? You don't need the entirety of the DPF, and you could just use the services that you require, e.g. for tracking a transfer process. You would need to write code to provision an endpoint access token using the Data Plane Signalling API.

**Reply by klumbe** [reply]
Sorry, didn't mention it before.
I thought it would be better to have it in one component, since I already have a data plane that fetches the files and pushes them to the endpoint (http-streaming sample).
It sounded more practical to already have some of these features in place and don't have to offer yet another port.

But if the framework does not support the use case I have in mind, I might be better of with the way that you suppose.

Not sure how much effort it would be to implement callbacks into the data plane, but I could imagine that the consumer data plane could simply act as HTTP proxy and the client-application is informed when a new file is ready at the provider.
Might be even more flexible. Still, the callback would most probably have to be send to the client application (ideally it the data plane would forward it).

**Reply by jimmarino** [reply]
You can run it in the same instance put use a different context to separate it from the internal API. Just create an extension with a JAX-RS controller.

**Reply by klumbe** [reply]
Many thanks for all the input.
Do you maybe have a link to some code to see how the integration with the signaling API works?
Read the docs, but it is rather theoretical in regards to register a data plane at the control plane and make sure that tokens are generated / destroyed upon state changes of the transfer process. Would love to see this in real code, just in case I'd need it.

**Reply by klumbe** [reply]
@jimmarino: 
Do you maybe have an example at hand that showcases how the token could be generated and invalidated based on the state of the `TransferProcess`?

As far as I can see, the `DataPlaneAuthorizationService` handles token storage and authorization.
I could imagine that my extension has to inject it and then use the `authorize` method with the token provided by the request,
but I don't see how the proper token could be identified if the pure JAX-RS request is not linked to the `TransferProcess`.
I also don't see how the generation and invalidation works in the push-scenario.
Does it mean the `TransferRequest` messages needs to be monitored in the control plane and the DSP message sent to the data provider be altered to contain the token?

Or is it not meant to be used in a custom API and it should be handled by the application completely such that the `TransferRequest` should already contain the authorization token in some way? That would be the least-preferred way, since it complicates invalidation based on the state of the `TransferProcess`.

If you are aware of any similar implementation and could point me into the right direction, I would really appreciate it.

**Reply by jimmarino** [reply]
Take a look at this:

https://eclipse-edc.github.io/documentation/for-contributors/data-plane/data-plane-signaling/

Let me know if that answers your questions or you have additional ones.

**Reply by klumbe** [reply]
Hi @jimmarino,
I've read that again, but I could not find any hint for the **PUSH** scenario.

I am faced with the following issues:
- `DataFlowStartMessage` is probably only sent to the providers data plane, since the `TransferRequest` does not contain any information about a "target data plane", just the `dataDestination` of `type` "HttpData" and the `baseUrl` which points at the data consumer's REST API.
This is a similar issue like the one I described when I thought that I could use a custom data sink for this purpose. I would need a way to only tell the receiving part that this sink is to be used when data is received by the public API. In case of a custom API, that is encoded in the `baseUrl`, but it still won't trigger notification of that custom API's data plane to generate the API-key and use it to send the actual DSP request to the data provider.
- How would a manually generated token be properly transmitted to the data provider? Is it a special field within the `dataDestination`?
I think that would not make any sense, since the request will only put the transfer-process into the "STARTING" state, but the token should be generated when it enters "STARTED" and invalidated when "SUSPENDED"
- How can the data plane with the custom API be informed that a transfer-process has been requested and that a token for the PUSH scenario needs to be generated?

So in general, the pure generation and validation of a token is not the issue, but I don't see how it can be coupled with a `TransferRequest` to create and destroy the token based on the transfer-process state.

Any hint would be appreciated.

**Reply by jimmarino** [reply]
It's difficult to follow your scenario because I can't discern which "data plane" and sinks you are referring to.

For push scenarios, the consumer control plane performs provisioning and creates a`DataAddress` that is included with the request to the provider. The `DataAddress` may contain a secret token. This `DataAddress` is passed from the provider control plane to the provider data plane along with the secret token. The data address must encode enough information so the provider data plane knows which endpoint to send data to and which wire protocol it should use. If a secret token is passed, that should provide enough correlation for the consumer endpoint (consumer data plane) to route the data.

**Reply by klumbe** [reply]
Hi @jimmarino,

my reference to the sink was meant to be on the consumer end, in particular in the push-scenario.
I thought it would be a good idea to have a reference to the sink the `TransferRequest` such that when it is created, the consumer would know where to pass the data, regardless of the chosen transfer type.

But I've switched to your suggestion to use a custom public HTTP API to receive the pushed data.
So it is an extension registering a Jersey controller offering a POST request on a dedicated port.

What you say sounds contrary to the statement in the documentation you have linked:

> When the control plane signals to the data plane to start a client pull transfer process, the data plane returns a DataAddress. This is only true for consumer-pull transfers - provider push transfers do not return a DataAddress

So my question would be:
**If that `DataAddress` and the token are generated in the consumer control plane in the PUSH-scenario, how could I access it from the custom API?**

My previous question around selecting the proper data plane - which may have caused the confusion - is related to the selection process in the consumers control plane: The doc states that the `DataFlowStartMessage` will be send to the appropriate data plane.
So I am wondering how the control plane could do this decision if the request itself contains just the `baseURL` of the `dataDestination`.
In my case, I will keep it as simple as possible and will just have a single data plane where the public API will be running in.
But theoretically, that custom API could be anything, like a dedicated web-server and in such a scenario, I would have to exchange the token, i.e. through a database.

Example of a `TransferRequest` sent to the consumer:
```json
{
    "@context": {
        "@vocab": "https://w3id.org/edc/v0.0.1/ns/"
    },
    "@type": "TransferRequest",
    "dataDestination": {
        "type": "HttpData",
        "baseUrl": "http://example.org:8080/public/custom/store/<id>}"
    },
    "protocol": "dataspace-protocol-http",
    "contractId": "...",
    "connectorId": "<provider-id>",
    "counterPartyAddress":  "<provider-dsp-url>/api/dsp",
    "transferType": "HttpData-PUSH"
}
```

For my scenario, the only thing of importance is that a token is generated, sent to the data provider and the token can be retrieved in the custom API to validate its correctness.
Since the doc actually states, that it is not generated in the PUSH scenario, I would guess that I would have to subscribe to state-changes in the consumers control plane and either generate or destroy a token. In the generate case, it should be part of DSP request forwarded to the provider and also in a `DataFlowStartMessage` shared with the data plane.
If I have to do this manually, I would need to know how to influence the DSP request and which particular fields are used.

Again, many thanks for your answer. I hope I have made it a bit more clear what I am actually missing.
Let me know if you need any further input.

**Reply by jimmarino** [reply]
Please look at how the `TransferProcessManager` and `Provisioner` (particularly the `provision` method) work on the consumer side where a token can be created, which is then passed to the provider in the `DataAddress` that is part of the transfer process request.

**Reply by klumbe** [reply]
@jimmarino: Thanks for that input. I will check these and will give some feedback as soon as I went through it.

Earlier today, I did some tcpdump and checked the messages flowing through DSP. The `dataAddress` was basically a copy of the `dataDestination` field in the `TransferRequest`, not containing any token.
In the [DataAddress]{https://github.com/eclipse-edc/Connector/blob/main/spi/common/core-spi/src/main/java/org/eclipse/edc/spi/types/domain/DataAddress.java} class, there is a field called `secret`. Would it be the one to populate for the token?

**Reply by klumbe** [reply]
I guess that the missing else branch in the [HttpProvisionerExtension](https://github.com/eclipse-edc/Connector/blob/main/extensions/control-plane/provision/provision-http/src/main/java/org/eclipse/edc/connector/controlplane/provision/http/HttpProvisionerExtension.java)  would be what I am actually searching for:
```java

            if (configuration.getProvisionerType() == ProvisionerConfiguration.ProvisionerType.PROVIDER) {
                var generator = new HttpProviderResourceDefinitionGenerator(configuration.getDataAddressType());
                manifestGenerator.registerGenerator(generator);
                monitor.debug(format("Registering provider provisioner: %s [%s]", configuration.getName(), configuration.getEndpoint().toString()));
            } else {
                monitor.warning(format("Client-side provisioning not yet supported by the %s. Skipping configuration for %s", name(), configuration.getName()));
            }
```

The class [AbstractHttpResourceDefinition](https://github.com/eclipse-edc/Connector/blob/main/extensions/control-plane/provision/provision-http/src/main/java/org/eclipse/edc/connector/controlplane/provision/http/impl/AbstractHttpResourceDefinition.java) gives the hint that the `dataAddressType` is used to determine the correct `Provisioner` (not too surprising).
Would this mean that I have to define a custom `dataAddressType` and build an Extension like the `HttpProvisionerExtension` to be able to fire the custom `Provisioner`?

Is it related to the `dataDestination.type` in the `TransferRequest`?
I am wondering if I could still use the provided `HttpData-PUSH` together with a custom type such that it would be clear that this processing is only required on the consumer end and only `dataDestination` is really that custom API.

**Reply by klumbe** [reply]
Hi @jimmarino,

as far as I understood, the `HttpProvisionerExtension` is handling provisioning requests against a HTTP callback URL, so that is not necessarily required here and I should build a custom `Provisioner`.

I tried that and on debugging, I found out that the `process.getResourcesToProvision()` method returns an empty list.
I couldn't find out all the details about how the `Provisioner` actually works, but I guess I have to define the resources requiring provisioning somewhere.

Scanning the code, this may be related to the `provision.manifest.verify` briefly mentioned as a policy scope in the old [Developer's Handbook](https://github.com/eclipse-edc/docs/blob/8089975980ebeb7f66c667b5a1654b6802d28df8/developer/handbook.md).
Does it mean a dedicated policy definition is required for running the provisioning, or is it just for evaluation during the provisioning itself?
Can't find it in the contract definition api, so I guess this is just outdated information.

Could someone give me an example how it should look like or point me to where I can find documentation on how to use it properly?

**Comment by jimmarino** [comment]
A `DataAddress` is just a pointer to data (i.e. an endpoint where it can be retrieved). There are internal data addresses associated with assets the connector uses to retrieve and send data associated with an asset. There are also external `DataAddresses` that can be sent to a counter-party to push or pull data.

The HTTP Provisioner provides an example of how to write a custom provisioner. The code is also well-documented and explains the relationship between provisioners and resource generators. You can also take a look at [S3BucketProvisioner](https://github.com/eclipse-edc/Technology-Aws/blob/1ea876837e266fa70ab4ce21029744dc6e134141/extensions/control-plane/provision/provision-aws-s3/src/main/java/org/eclipse/edc/connector/provision/aws/s3/S3BucketProvisioner.java#L37) to see how to generate and send access tokens.
